manager/controllers/app/select_modules.go

```diff
@@ -9,70 +9,47 @@ import (
-	Context *app.DataContext
-	// Evaluated config policies
-	Configuration adminconfig.EvaluatorOutput
-	// Workload cluster
-	WorkloadCluster multicluster.Cluster
-	// Governance actions to perform on this asset
-	Actions []taxonomy.Action
-}
-
 // Temporary hard-coded capability representing Actions of read/copy capability.
 // A change to modules is requested to add "transform" as an additional capability to the existing read and copy modules.
 const Transform = "transform"
 
-// Node represents an access point to data (as a physical source/sink, or a virtual endpoint)
-// A virtual endpoint is activated by the workload for read/write actions.
-type Node struct {
-	Connection *taxonomy.Interface
-	Virtual    bool
-}
-
-// Edge represents a module capability that gets data via source and returns data via sink interface
-type Edge struct {
-	Source          *Node
-	Sink            *Node
-	Module          *app.FybrikModule
-	CapabilityIndex int
+// component responsible for data path construction
+type PathBuilder struct {
+	Log   *zerolog.Logger
+	Env   *Environment
+	Asset *DataInfo
 }
 
-// ResolvedEdge extends an Edge by adding actions that a module should perform, and the cluster where the module will be deployed
-// TODO(shlomitk1): add plugins/transformation capabilities to this structure
-type ResolvedEdge struct {
-	Edge
-	Actions        []taxonomy.Action
-	Cluster        string
-	StorageAccount app.FybrikStorageAccountSpec
-}
-
-// Solution is a final solution enabling a plotter construction.
-// It represents a full data flow between the data source and the workload.
-type Solution struct {
-	DataPath []*ResolvedEdge
+// find a solution for data plane orchestration
+func (p *PathBuilder) solve() (Solution, error) {
+	p.Log.Trace().Str(logging.DATASETID, p.Asset.Context.DataSetID).Msg("Choose modules for dataset")
+	solutions := p.FindPaths(p.Asset)
+	// No data path found for the asset
+	if len(solutions) == 0 {
+		msg := "Deployed modules do not provide the functionality required to construct a data path"
+		p.Log.Error().Str(logging.DATASETID, p.Asset.Context.DataSetID).Msg(msg)
+		logging.LogStructure("Data Item Context", p.Asset, p.Log, true, true)
+		logging.LogStructure("Module Map", p.Env.Modules, p.Log, true, true)
+		return Solution{}, errors.New(msg + " for " + p.Asset.Context.DataSetID)
+	}
+	return solutions[0], nil
 }
 
 // FindPaths finds all valid data paths between the data source and the workload
 // First, data paths are constructed using interface connections, starting from data source.
 // Then, transformations are added to the found paths, and clusters are matched to satisfy restrictions from admin config policies.
 // Optimization is done by the shortest path (the paths are sorted by the length). To be changed in future versions.
-func (p *PlotterGenerator) FindPaths(item *DataInfo, appContext *app.FybrikApplication) []Solution {
+func (p *PathBuilder) FindPaths(item *DataInfo) []Solution {
```

Comment:
Why does `FindPaths()` need a `DataInfo` parameter? It can just access `p.Asset`, can't it?
(same for other methods of `PathBuilder`)

Reply:
It doesn't, forgot to remove...